next: DropdownMenu docs (#632)
huntabyte/bits-ui: sites/docs/content/components/dropdown-menu.md

</ComponentPreviewV2>

## Structure

```svelte
<script lang="ts">
	import { DropdownMenu } from "bits-ui";
</script>

<DropdownMenu.Root>
	<DropdownMenu.Trigger />

	<DropdownMenu.Content>

> 		<DropdownMenu.Label />

Reviewer: ```suggestion
		<DropdownMenu.Group>
		<DropdownMenu.GroupLabel />
```

In previous versions of this, we weren't appropriately requiring a label to be associated with a group, which is confusing. In Svelte 5, if you're using a `GroupLabel`, it must be within either a `DropdownMenu.Group` or `DropdownMenu.RadioGroup`.

		</DropdownMenu.RadioGroup>

		<DropdownMenu.Sub>
			<DropdownMenu.SubTrigger />
			<DropdownMenu.SubContent />
		</DropdownMenu.Sub>

		<DropdownMenu.Separator />
		<DropdownMenu.Arrow />
	</DropdownMenu.Content>
</DropdownMenu.Root>
```

## Reusable Components

If you're planning to use Dropdown Menu in multiple places, you can create a reusable component that wraps the Dropdown Menu component.

This example shows you how to create a Dropdown Menu component that accepts a few custom props that make it more capable.

```svelte title="CustomDropdownMenu.svelte"
<script lang="ts">
	import type { Snippet } from "svelte";
	import { DropdownMenu, type WithoutChild } from "bits-ui";

	type Props = DropdownMenu.Props & {
		buttonText: string;
		items: { label: string; value: string }[];
		contentProps?: WithoutChild<DropdownMenu.Content.Props>;
		// other component props if needed
	};

	let {
		open = $bindable(false),
		children,
		buttonText,
		items,
		contentProps,
		...restProps
	}: Props = $props();
</script>

<DropdownMenu.Root bind:open {...restProps}>
	<DropdownMenu.Trigger>
		{buttonText}
	</DropdownMenu.Trigger>
	<DropdownMenu.Content {...contentProps}>

> 		<DropdownMenu.Group>

Reviewer: We should add a label here or at least expose an `aria-label` property to apply to the group 😃

Author: Went with an `aria-label` for the sole purpose of showcasing that there is a different alternative 🤷, as I wasn't sure if repeating information visually (the `buttonText` and now `Label`) was worth it.

			{#each items as item}

> 				<DropdownMenu.Item textValue={item.value}>

Reviewer: In this case `textValue === item.label`. `textValue` is only necessary if you have some complex content within the item and is only used for typeahead purposes

Author: I gave this a minute, and since there is no kind of `search` functionality going on here, wouldn't it be better to forego the `label`/`value` duo and just pass the values themselves. WDYT?

> 				<DropdownMenu.Item textValue={item.value}>

Reviewer: ```suggestion
				<DropdownMenu.Item textValue={item.label}>
```

### Two-Way Binding

Use the `bind:open` directive for effortless two-way synchronization between your local state and the menu's internal state.

```svelte {3,6,8}
<script lang="ts">
	import { DropdownMenu } from "bits-ui";
	let isOpen = $state(false);
</script>

<button on:click={() => (isOpen = true)}>Open Dropdown Menu</button>

<DropdownMenu.Root bind:open={isOpen}>
	<!-- Dropdown Menu content -->
</DropdownMenu.Root>
```

This setup enables opening the menu via the custom button and ensures the local `isOpen` state updates when the menu closes through any means (e.g., escape key).

### Change Handler

You can also use the `onOpenChange` prop to update local state when the menu's `open` state changes. This is useful when you don't want two-way binding for one reason or another, or you want to perform additional logic when the menu opens or closes.

```svelte {3,7-11}
<script lang="ts">
	import { DropdownMenu } from "bits-ui";
	let isOpen = $state(false);
</script>

<DropdownMenu.Root
	open={isOpen}
	onOpenChange={(open) => {
		isOpen = open;
		// additional logic here.
	}}
>
	<!-- ... -->
</DropdownMenu.Root>
```

## Nested Menus

> You can nest menus within menus to create complex menu structures.

Reviewer: ```suggestion
You can create nested menus using the `DropdownMenu.Sub` component to create complex menu structures.
```

> ```svelte

Reviewer: ```suggestion
```svelte /DropdownMenu.Sub/
```
